src/backend/mpi/coherence.cpp

```diff
@@ -0,0 +1,208 @@
+/**
+ * @brief A cache mutex protects all operations on cacheControl
+ * @deprecated Should eventually be handled by a cache module
+ */
+extern pthread_mutex_t cachemutex;
+/**
+ * @brief ibsem is used to serialize all Infiniband (MPI) operations
+ * @deprecated Should not be needed once the cache module is implemented
+ */
+extern sem_t ibsem;
+/**
+ * @brief barwindowsused is needed to know which globalDataWindows
+ * to close.
+ * @deprecated No longer needed once proper API functions are in place
+ */
+extern char * barwindowsused;
+/**
+ * @brief globalDataWindows need to be explicitly closed by the
+ * caller of storepageDIFF
+ * @deprecated Should eventually be handled through API functions
+ */
+extern MPI_Win *globalDataWindow;
+/**
+ * @brief sharerWindow protects the pyxis directory
+ * @deprecated Should not be needed once the pyxis directory is
+ * managed from elsewhere through a cache module.
+ */
+extern MPI_Win sharerWindow;
+/**
+ * @brief Needed to update argo statistics
+ * @deprecated Should be replaced by API calls to a stats module
+ */
+extern argo_statistics stats;
+/**
+ * @brief Needed to update information about cache pages touched
+ * @deprecated Should eventually be handled by a cache module
+ */
+extern argo_byte *touchedcache;
+/**
+ * @brief workcomm is needed to poke the MPI system during one sided RMA
+ */
+extern MPI_Comm workcomm;
+/**
+ * @brief Write buffer to ensure selectively handled pages can be removed
+ * @deprecated This should eventually be handled by a cache module
+ */
+extern write_buffer<std::size_t>* argo_write_buffer;
+
+/** @brief The size of an ArgoDSM page */
+const std::size_t pagesize = GRAN;
+
+namespace argo {
+	namespace backend {
+		void selective_acquire(void *addr, size_t size){
+			if(size == 0){
+				// Nothing to invalidate
+				return;
+			}
+
+			const std::size_t block_size = pagesize*CACHELINE;
+			const std::size_t start_address = reinterpret_cast<std::size_t>(argo::virtual_memory::start_address());
+			const std::size_t page_misalignment = reinterpret_cast<std::size_t>(addr)%block_size;
+			std::size_t argo_address =
+				((reinterpret_cast<std::size_t>(addr)-start_address)/block_size)*block_size;
+			const std::size_t node_id = argo::backend::node_id();
+			const std::size_t node_id_bit = 1 << node_id;
+
+			// Lock relevant mutexes. Start statistics timekeeping
+			double t1 = MPI_Wtime();
+			pthread_mutex_lock(&cachemutex);
+			sem_wait(&ibsem);
+
+			// Iterate over all pages to selectively invalidate
+			for(std::size_t page_address = argo_address;
+					page_address < argo_address + page_misalignment + size;
+					page_address += block_size){
+				const std::size_t cache_index = getCacheIndex(page_address);
+				const std::size_t classification_index = get_classification_index(page_address);
+
+				// If the page is dirty, downgrade it
+				if(cacheControl[cache_index].dirty == DIRTY){
+					argo_write_buffer->erase(cache_index);
```

Comment:
It is currently safe placed both before and after the writeback as `ibsem` and `cachemutex` currently prevent simultaneous operations from happening on both the cache and the write buffer.

I also believe that for any foreseeable future, simultaneous cache operations on the same cached page will not be permitted (even in the parallel backend I lock access to at least each cache index separately), and therefore it should be safe to perform these in any order so long as there is mutual exclusion on the cache index.

However, as the cache is not set to `CLEAN` or `INVALID` until after `storepageDIFF` has been performed in `swdsm.cpp`, it may be preferable to keep this behaviour uniform and erase the write buffer entry after creating the diff if only to keep the order of operations the same. I will make this change in the next push.

Reply:
Both setting the cache to `CLEAN` and removing the cache index from the write buffer now occurs _after_ the call to `storepageDIFF`.